Mage.Sets/src/mage/cards/d/DesertWarfare.java

```diff
@@ -0,0 +1,200 @@
+
+        this.addAbility(new PutCardIntoGraveFromHandLibraryAllTriggeredAbility(Zone.BATTLEFIELD, effect,
+                false, TargetController.YOU, SetTargetPointer.CARD).setTriggerPhrase("Whenever you sacrifice a Desert and whenever a Desert card is put into your graveyard from your hand or library, put that card onto the battlefield under your control at the beginning of your next end step."));
+
+        //At the beginning of combat on your turn, if you control five or more Deserts, create that many 1/1 red, green, and white Sand Warrior creature tokens. They gain haste.
+        //Based on Palani's Hatcher
+        this.addAbility(new ConditionalInterveningIfTriggeredAbility(
+                new BeginningOfCombatTriggeredAbility(new DesertWarfareCreateTokensEffect(), TargetController.YOU, false),
+                new PermanentsOnTheBattlefieldCondition(filter, ComparisonType.OR_GREATER, 5, true),
+                "At the beginning of combat on your turn, if you control five or more Deserts, create that many 1/1 red, green, and white Sand Warrior creature tokens. They gain haste.").addHint(DesertsYouControlHint.instance));
+
+    }
+
+    private DesertWarfare(final DesertWarfare card) {
+        super(card);
+    }
+
+    @Override
+    public DesertWarfare copy() {
+        return new DesertWarfare(this);
+    }
+
+}
+
+//Based on PutCardIntoGraveFromAnywhereAllTriggeredAbility
+class PutCardIntoGraveFromHandLibraryAllTriggeredAbility extends TriggeredAbilityImpl {
+
+    private static FilterCard filterCard = new FilterCard();
+
+    static {
+        filterCard.add(SubType.DESERT.getPredicate());
+    }
+
+    private final SetTargetPointer setTargetPointer;
+
+    public PutCardIntoGraveFromHandLibraryAllTriggeredAbility(Zone zone, Effect effect, boolean optional, TargetController targetController, SetTargetPointer setTargetPointer) {
+        super(zone, effect, optional);
+        FilterCard filter = filterCard.copy();
+        this.setTargetPointer = setTargetPointer;
+        filter.add(targetController.getOwnerPredicate());
+    }
+
+    protected PutCardIntoGraveFromHandLibraryAllTriggeredAbility(final PutCardIntoGraveFromHandLibraryAllTriggeredAbility ability) {
+        super(ability);
+        filterCard = filterCard.copy();
+        this.setTargetPointer = ability.setTargetPointer;
+    }
+
+    @Override
+    public PutCardIntoGraveFromHandLibraryAllTriggeredAbility copy() {
+        return new PutCardIntoGraveFromHandLibraryAllTriggeredAbility(this);
+    }
+
+    @Override
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.ZONE_CHANGE;
+    }
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        Card card = game.getCard(event.getTargetId());
+        if (((ZoneChangeEvent) event).getToZone() == Zone.GRAVEYARD
```

Comment:
don't repeat checks in both branches of an if-else, perform the common checks first. return false at everything that doesn't trigger, and only have a single setTargetPointer at the end before returning true

also it's simpler to declare `ZoneChangeEvent zEvent = (ZoneChangeEvent) event` so you only have to cast once